File: app/src/main/java/com/najmi/oreamnos/LogListActivity.java
package com.najmi.oreamnos;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.app.AppCompatDelegate;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.appbar.MaterialToolbar;
import com.google.android.material.button.MaterialButton;
import com.najmi.oreamnos.model.UsageStats;
import com.najmi.oreamnos.utils.PreferencesManager;

import java.util.List;
import java.util.Locale;

/**
 * Activity for displaying all logs in a dedicated full-screen list.
 */
public class LogListActivity extends AppCompatActivity {

    private PreferencesManager prefsManager;
    private RecyclerView logsRecyclerView;
    private TextView emptyLogsText;
    private TextView logCountText;
    private MaterialButton clearLogsButton;
    private LogAdapter logAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        prefsManager = new PreferencesManager(this);
        applyTheme(prefsManager.getTheme());

        setContentView(R.layout.activity_log_list);

        // Setup toolbar
        MaterialToolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        toolbar.setNavigationOnClickListener(v -> finish());

        initViews();
        loadLogs();
    }

    private void initViews() {
        logsRecyclerView = findViewById(R.id.logsRecyclerView);
        emptyLogsText = findViewById(R.id.emptyLogsText);
        logCountText = findViewById(R.id.logCountText);
        clearLogsButton = findViewById(R.id.clearLogsButton);

        logsRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        logAdapter = new LogAdapter();
        logAdapter.setOnLogClickListener(this::showLogDetailsDialog);
        logsRecyclerView.setAdapter(logAdapter);

        clearLogsButton.setOnClickListener(v -> {
            new AlertDialog.Builder(this)
                    .setTitle(R.string.clear_logs)
                    .setMessage("Are you sure you want to clear all logs?")
                    .setPositiveButton("Clear", (dialog, which) -> {
                        prefsManager.clearLogs();
                        loadLogs();
                        Toast.makeText(this, R.string.logs_cleared, Toast.LENGTH_SHORT).show();
                    })
                    .setNegativeButton("Cancel", null)
                    .show();
        });
    }

    private void loadLogs() {
        UsageStats stats = prefsManager.getUsageStats();
        List<UsageStats.LogEntry> logs = stats.getLogs();

        int count = logs != null ? logs.size() : 0;
        logCountText.setText(String.format(Locale.US, "%d log entries", count));

        if (logs == null || logs.isEmpty()) {
            emptyLogsText.setVisibility(View.VISIBLE);
            logsRecyclerView.setVisibility(View.GONE);
        } else {
            emptyLogsText.setVisibility(View.GONE);
            logsRecyclerView.setVisibility(View.VISIBLE);
            logAdapter.setLogs(logs);
        }
    }

    private void showLogDetailsDialog(UsageStats.LogEntry log) {
        View dialogView = LayoutInflater.from(this).inflate(R.layout.dialog_log_details, null);

        TextView levelBadge = dialogView.findViewById(R.id.dialogLevelBadge);
        TextView tagText = dialogView.findViewById(R.id.dialogTagText);
        TextView timeText = dialogView.findViewById(R.id.dialogTimeText);
        TextView messageText = dialogView.findViewById(R.id.dialogMessageText);
        TextView detailsText = dialogView.findViewById(R.id.dialogDetailsText);
        MaterialButton copyButton = dialogView.findViewById(R.id.dialogCopyButton);

        levelBadge.setText(log.getLevel());

        int badgeRes;
        switch (log.getLevel()) {
            case UsageStats.LogEntry.LEVEL_ERROR:
                badgeRes = R.drawable.log_badge_error;
                break;
            case UsageStats.LogEntry.LEVEL_WARNING:
                badgeRes = R.drawable.log_badge_warn;
                break;
            case UsageStats.LogEntry.LEVEL_DEBUG:
                badgeRes = R.drawable.log_badge_debug;
                break;
            case UsageStats.LogEntry.LEVEL_INFO:
            default:
                badgeRes = R.drawable.log_badge_info;
                break;
        }
        levelBadge.setBackgroundResource(badgeRes);

        tagText.setText(log.getTag() != null ? log.getTag() : "App");
        timeText.setText(log.getFormattedDate());
        messageText.setText(log.getMessage());

        String details = log.getDetails();
        if (details != null && !details.isEmpty()) {
            detailsText.setText(details);
        } else {
            detailsText.setText("No additional details.");
            detailsText.setTypeface(null, android.graphics.Typeface.ITALIC);
        }

        AlertDialog dialog = new com.google.android.material.dialog.MaterialAlertDialogBuilder(this)
                .setTitle("Log Details")
                .setView(dialogView)
                .setPositiveButton("Close", null)
                .create();

        copyButton.setOnClickListener(v -> {
            StringBuilder sb = new StringBuilder();
            sb.append("Level: ").append(log.getLevel()).append("\n");
            sb.append("Time: ").append(log.getFormattedDate()).append("\n");
            sb.append("Tag: ").append(log.getTag()).append("\n");
            sb.append("Message: ").append(log.getMessage()).append("\n");
            if (log.getDetails() != null) {
                sb.append("Details:\n").append(log.getDetails());
            }

            ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
            ClipData clip = ClipData.newPlainText("Log Details", sb.toString());
            clipboard.setPrimaryClip(clip);
            Toast.makeText(this, "Log details copied", Toast.LENGTH_SHORT).show();
        });

        dialog.show();
    }

    private void applyTheme(String theme) {
        int mode;
        switch (theme) {
            case PreferencesManager.THEME_LIGHT:
                mode = AppCompatDelegate.MODE_NIGHT_NO;
                break;
            case PreferencesManager.THEME_DARK:
                mode = AppCompatDelegate.MODE_NIGHT_YES;
                break;
            case PreferencesManager.THEME_SYSTEM:
            default:
                mode = AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM;
                break;
        }
        AppCompatDelegate.setDefaultNightMode(mode);
    }

    // ==================== LOG ADAPTER ====================

    private static class LogAdapter extends RecyclerView.Adapter<LogAdapter.LogViewHolder> {

        private List<UsageStats.LogEntry> logs;
        private OnLogClickListener listener;

        public interface OnLogClickListener {
            void onLogClick(UsageStats.LogEntry log);
        }

        public void setLogs(List<UsageStats.LogEntry> logs) {
            this.logs = logs;
            notifyDataSetChanged();
        }

        public void setOnLogClickListener(OnLogClickListener listener) {
            this.listener = listener;
        }

        @NonNull
        @Override
        public LogViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.item_log_entry, parent, false);
            return new LogViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull LogViewHolder holder, int position) {
            if (logs == null || position >= logs.size())
                return;
            UsageStats.LogEntry log = logs.get(position);
            holder.bind(log);

            holder.itemView.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onLogClick(log);
                }
            });
        }

        @Override
        public int getItemCount() {
            return logs != null ? logs.size() : 0;
        }

        static class LogViewHolder extends RecyclerView.ViewHolder {
            private final TextView levelBadge;
            private final TextView tagText;
            private final TextView timeText;
            private final TextView messageText;

            LogViewHolder(@NonNull View itemView) {
                super(itemView);
                levelBadge = itemView.findViewById(R.id.levelBadge);
                tagText = itemView.findViewById(R.id.tagText);
                timeText = itemView.findViewById(R.id.timeText);
                messageText = itemView.findViewById(R.id.messageText);
            }

            void bind(UsageStats.LogEntry log) {
                String level = log.getLevel();
                levelBadge.setText(level);

                int badgeRes;
                switch (level) {
                    case UsageStats.LogEntry.LEVEL_ERROR:
                        badgeRes = R.drawable.log_badge_error;
                        break;
                    case UsageStats.LogEntry.LEVEL_WARNING:
                        badgeRes = R.drawable.log_badge_warn;
                        break;
                    case UsageStats.LogEntry.LEVEL_DEBUG:
                        badgeRes = R.drawable.log_badge_debug;
                        break;
                    case UsageStats.LogEntry.LEVEL_INFO:
                    default:
                        badgeRes = R.drawable.log_badge_info;
                        break;
                }
                levelBadge.setBackgroundResource(badgeRes);

                tagText.setText(log.getTag() != null ? log.getTag() : "App");
                timeText.setText(log.getFormattedTime());

                String msg = log.getMessage();
                String details = log.getDetails();
                if (details != null && !details.isEmpty()) {
                    messageText.setText(msg + ": " + details);
                } else {
                    messageText.setText(msg);
                }
            }
        }
    }
}

File: app/src/main/java/com/najmi/oreamnos/LogListActivity.kt
package com.najmi.oreamnos

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowBack
import androidx.compose.material.icons.filled.Delete
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.najmi.oreamnos.model.UsageStats
import com.najmi.oreamnos.ui.theme.ErrorRed
import com.najmi.oreamnos.ui.theme.SocurateTheme
import com.najmi.oreamnos.ui.theme.SuccessGreen
import com.najmi.oreamnos.utils.PreferencesManager

/**
 * Activity for displaying all logs in a dedicated full-screen list.
 * Converted to Jetpack Compose.
 */
class LogListActivity : ComponentActivity() {
    
    private lateinit var prefsManager: PreferencesManager
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        
        prefsManager = PreferencesManager(this)
        
        setContent {
            var logs by remember { mutableStateOf(prefsManager.getUsageStats().getLogs()) }
            val currentTheme = prefsManager.getTheme()
            
            SocurateTheme(themeMode = currentTheme) {
                LogListScreen(
                    logs = logs,
                    onNavigateBack = { finish() },
                    onClearLogs = {
                        prefsManager.clearLogs()
                        logs = emptyList()
                    }
                )
            }
        }
    }
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun LogListScreen(
    logs: List<UsageStats.LogEntry>,
    onNavigateBack: () -> Unit,
    onClearLogs: () -> Unit
) {
    val context = LocalContext.current
    var showClearDialog by remember { mutableStateOf(false) }
    var selectedLog by remember { mutableStateOf<UsageStats.LogEntry?>(null) }
    
    // Clear confirmation dialog
    if (showClearDialog) {
        AlertDialog(
            onDismissRequest = { showClearDialog = false },
            title = { Text("Clear Logs") },
            text = { Text("Are you sure you want to clear all logs?") },
            confirmButton = {
                TextButton(onClick = {
                    onClearLogs()
                    showClearDialog = false
                    Toast.makeText(context, "Logs cleared", Toast.LENGTH_SHORT).show()
                }) {
                    Text("Clear", color = MaterialTheme.colorScheme.error)
                }
            },
            dismissButton = {
                TextButton(onClick = { showClearDialog = false }) {
                    Text("Cancel")
                }
            }
        )
    }
    
    // Log details dialog
    selectedLog?.let { log ->
        LogDetailsDialog(
            log = log,
            onDismiss = { selectedLog = null },
            onCopy = {
                val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                val text = buildLogText(log)
                clipboard.setPrimaryClip(ClipData.newPlainText("Log Details", text))
                Toast.makeText(context, "Log details copied", Toast.LENGTH_SHORT).show()
            }
        )
    }
    
    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Log History") },
                navigationIcon = {
                    IconButton(onClick = onNavigateBack) {
                        Icon(
                            imageVector = Icons.AutoMirrored.Filled.ArrowBack,
                            contentDescription = "Back"
                        )
                    }
                },
                actions = {
                    Text(
                        text = "${logs.size} entries",
                        style = MaterialTheme.typography.bodyMedium,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        modifier = Modifier.padding(end = 8.dp)
                    )
                    IconButton(onClick = { showClearDialog = true }) {
                        Icon(
                            imageVector = Icons.Default.Delete,
                            contentDescription = "Clear Logs",
                            tint = MaterialTheme.colorScheme.error
                        )
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = MaterialTheme.colorScheme.background
                )
            )
        },
        containerColor = MaterialTheme.colorScheme.background
    ) { paddingValues ->
        if (logs.isEmpty()) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(paddingValues),
                contentAlignment = Alignment.Center
            ) {
                Text(
                    text = "No logs yet",
                    style = MaterialTheme.typography.bodyLarge,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    textAlign = TextAlign.Center
                )
            }
        } else {
            LazyColumn(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(paddingValues)
                    .padding(horizontal = 16.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                item { Spacer(modifier = Modifier.height(8.dp)) }
                
                items(logs) { log ->
                    LogItem(
                        log = log,
                        onClick = { selectedLog = log }
                    )
                }
                
                item { Spacer(modifier = Modifier.height(16.dp)) }
            }
        }
    }
}

@Composable
fun LogItem(
    log: UsageStats.LogEntry,
    onClick: () -> Unit
) {
    Surface(
        modifier = Modifier
            .fillMaxWidth()
            .clickable { onClick() },
        shape = RoundedCornerShape(12.dp),
        color = MaterialTheme.colorScheme.surface
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(12.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            // Level badge
            Surface(
                shape = RoundedCornerShape(4.dp),
                color = getLevelColor(log.level)
            ) {
                Text(
                    text = log.level ?: "INFO",
                    style = MaterialTheme.typography.labelSmall,
                    color = Color.White,
                    modifier = Modifier.padding(horizontal = 6.dp, vertical = 2.dp)
                )
            }
            
            Spacer(modifier = Modifier.width(12.dp))
            
            // Content
            Column(modifier = Modifier.weight(1f)) {
                Row(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween
                ) {
                    Text(
                        text = log.tag ?: "App",
                        style = MaterialTheme.typography.labelMedium,
                        color = MaterialTheme.colorScheme.primary
                    )
                    Text(
                        text = log.formattedTime,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
                Spacer(modifier = Modifier.height(4.dp))
                Text(
                    text = buildLogPreview(log),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurface,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis
                )
            }
        }
    }
}

@Composable
fun LogDetailsDialog(
    log: UsageStats.LogEntry,
    onDismiss: () -> Unit,
    onCopy: () -> Unit
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Log Details") },
        text = {
            Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
                Row {
                    Surface(
                        shape = RoundedCornerShape(4.dp),
                        color = getLevelColor(log.level)
                    ) {
                        Text(
                            text = log.level ?: "INFO",
                            style = MaterialTheme.typography.labelSmall,
                            color = Color.White,
                            modifier = Modifier.padding(horizontal = 6.dp, vertical = 2.dp)
                        )
                    }
                    Spacer(modifier = Modifier.width(8.dp))
                    Text(
                        text = log.tag ?: "App",
                        style = MaterialTheme.typography.labelMedium,
                        color = MaterialTheme.colorScheme.primary
                    )
                }
                
                Text(
                    text = log.formattedDate,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
                
                Text(
                    text = log.message ?: "",
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurface
                )
                
                if (!log.details.isNullOrEmpty()) {
                    Text(
                        text = log.details!!,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                } else {
                    Text(
                        text = "No additional details.",
                        style = MaterialTheme.typography.bodySmall.copy(fontStyle = FontStyle.Italic),
                        color = MaterialTheme.colorScheme.onSurfaceVariant
                    )
                }
            }
        },
        confirmButton = {
            TextButton(onClick = onCopy) {
                Text("Copy")
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) {
                Text("Close")
            }
        }
    )
}

private fun getLevelColor(level: String?): Color {
    return when (level) {
        UsageStats.LogEntry.LEVEL_ERROR -> ErrorRed
        UsageStats.LogEntry.LEVEL_WARNING -> Color(0xFFFBBC05)  // Yellow
        UsageStats.LogEntry.LEVEL_DEBUG -> Color(0xFF9E9E9E)   // Gray
        else -> Color(0xFF4CAF50)  // Green for INFO
    }
}

private fun buildLogPreview(log: UsageStats.LogEntry): String {
    val msg = log.message ?: ""
    val details = log.details
    return if (!details.isNullOrEmpty()) "$msg: $details" else msg
}

private fun buildLogText(log: UsageStats.LogEntry): String {
    return buildString {
        append("Level: ${log.level}\n")
        append("Time: ${log.formattedDate}\n")
        append("Tag: ${log.tag}\n")
        append("Message: ${log.message}\n")
        if (!log.details.isNullOrEmpty()) {
            append("Details:\n${log.details}")
        }
    }
}
